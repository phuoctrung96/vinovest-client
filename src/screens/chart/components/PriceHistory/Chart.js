import PropTypes from "prop-types";
import React from "react";
import { LineChart, Line, XAxis, YAxis, CartesianGrid, Tooltip, ResponsiveContainer } from "recharts";
import { useTheme } from "#shared/hooks";
import { Tooltip as StyledTooltip } from "./components/styled";
import moment from "moment";
import get from "lodash/get";
import { useTranslation } from "react-i18next";

const DATE_PATTERN = "MMM YYYY";

const getDate = (date) => moment(new Date(date)).format(DATE_PATTERN);

const mockData = [
    {
        date: getDate("2017-09-03"),
        currentWine: 4000,
        benchmark: 3000,
    },
    {
        date: getDate("2018-09-03"),
        currentWine: 4020,
        benchmark: 1000,
    },
    {
        date: getDate("2019-09-03"),
        currentWine: 4500,
        benchmark: 3800,
    },
    {
        date: getDate("2016-09-03"),
        currentWine: 2000,
        benchmark: 1500,
    },
    {
        date: getDate("2015-09-03"),
        currentWine: 4200,
        benchmark: 1000,
    },
    {
        date: getDate("2014-09-03"),
        currentWine: 3000,
        benchmark: 2000,
    },
];

const CustomTooltip = ({ active, payload }) => {
    const { t } = useTranslation(["chart"]);
    const data = get(payload, ["0", "payload"]);
    const key = get(payload, ["0", "dataKey"]);
    if (active && data) {
        const { date } = data;
        const price = data[key];

        return (
            <StyledTooltip>
                <span className="date">{date}</span>
                <p className="price">
                    Average price:
                    <strong>${price.toLocaleString()}</strong>
                </p>
            </StyledTooltip>
        );
    }

    return null;
};

const AxisTick = (props) => {
    const { x, y, payload } = props;
    const date = moment(payload.value, DATE_PATTERN);
    return (
        <g transform={`translate(${x},${y})`} height={150}>
            <text x="0" y="0" height={120} fill="#828282" textAnchor="middle">
                <tspan x="0" dy="1.2em" fontSize={12}>
                    {date.format("DD MMM")}
                </tspan>
                <tspan x="0" dy="1.2em" fontSize={10}>
                    {date.format("YYYY")}
                </tspan>
            </text>
        </g>
    );
};

const YAxisTick = (props) => {
    const { x, y, payload } = props;
    return (
        <g transform={`translate(${x},${y})`} height={150}>
            <text x="-15" y="0" height={120} fill="#828282" fontWeight={500} fontSize={12} textAnchor="middle">
                {payload.value.toLocaleString("en-US", {
                    style: "currency",
                    currency: "USD",
                    maximumSignificantDigits: 1,
                })}
            </text>
        </g>
    );
};
/**
 *
 * @returns {*}
 * @constructor
 */
const Chart = ({ showBenchmark, data = mockData }) => {
    const color = useTheme("colors");
    return (
        <ResponsiveContainer width="100%" height={300}>
            <LineChart width="100%" height={300} data={data}>
                <CartesianGrid stroke="#BDBDBD" horizontal={false} strokeDasharray="0" />
                <XAxis
                    stroke="#F1F1F1"
                    dataKey="date"
                    tick={AxisTick}
                    tickFormatter={(time) => {
                        return moment(time, DATE_PATTERN).format("DD ".concat(DATE_PATTERN));
                    }}
                />
                <YAxis stroke="#F1F1F1" tick={YAxisTick} />
                <Tooltip content={<CustomTooltip />} />
                <Line
                    type="monotone"
                    dataKey="currentWine"
                    dot={false}
                    stroke={color.green}
                    activeDot={{ stroke: color.black, fill: color.green, strokeWidth: 4, r: 10 }}
                />
                {showBenchmark && (
                    <Line type="monotone" dataKey="benchmark" activeDot={false} dot={false} stroke="#BDBDBD" />
                )}
            </LineChart>
        </ResponsiveContainer>
    );
};

Chart.defaultProps = {
    data: mockData,
    showBenchmark: true,
};

Chart.propTypes = {
    data: PropTypes.arrayOf(PropTypes.object),
    showBenchmark: PropTypes.bool,
};

export default Chart;
