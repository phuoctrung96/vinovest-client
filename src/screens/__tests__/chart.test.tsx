import React from "react";
import { render } from "@testing-library/react";
import Chart from "../chart/ChartScreen";
import { withContexts } from "#shared/testing/wrappers";

jest.mock("@okta/okta-react", () => ({
    useOktaAuth: () => ({
        authState: { isAuthenticated: true },
        authService: { handleAuthentication: jest.fn() },
    }),
    withOktaAuth: (x) => x,
}));
describe("Chart component", () => {
    let component;
    const props = {
        portalURIFetch: {
            status: {
                isSuccess: jest.fn(() => true),
                isInvalidated: jest.fn(() => true),
            },
            data: "hello",
        },
    };
    const { component: Comp } = withContexts(Chart);
    component = Comp;

    it("Should match snapshot", () => {
        const { baseElement } = render(component(props));
        expect(baseElement).toMatchSnapshot();
    });
});
