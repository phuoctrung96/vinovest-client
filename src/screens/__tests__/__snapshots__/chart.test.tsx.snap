// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chart component Should match snapshot 1`] = `
.c9 {
  text-transform: uppercase;
  background-color: transparent;
  outline: none;
  border: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-weight: 500;
  font-size: 12px;
  line-height: 25px;
  -webkit-letter-spacing: 1px;
  -moz-letter-spacing: 1px;
  -ms-letter-spacing: 1px;
  letter-spacing: 1px;
  cursor: pointer;
}

.c5 {
  width: 13px;
  height: 13px;
  border-radius: 100%;
  background-color: #27AE60;
  margin-right: 18px;
}

.c8 {
  width: 13px;
  height: 13px;
  border-radius: 100%;
  background-color: #BDBDBD;
  margin-right: 18px;
}

.c6 {
  font-weight: 500;
  font-size: 18px;
  -webkit-letter-spacing: 0.005em;
  -moz-letter-spacing: 0.005em;
  -ms-letter-spacing: 0.005em;
  letter-spacing: 0.005em;
}

.c7 {
  font-weight: 500;
  font-size: 12px;
  -webkit-letter-spacing: 0.005em;
  -moz-letter-spacing: 0.005em;
  -ms-letter-spacing: 0.005em;
  letter-spacing: 0.005em;
  color: #bdbdbd;
  margin: 0 4px;
}

.c2 {
  height: 2px;
  background-color: #f1f1f1;
  width: 100%;
  margin-top: 12px;
  margin-bottom: 32px;
  padding: 0;
  border: none;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c4:not(:last-child) {
  margin-bottom: 16px;
}

.c0 {
  width: 90%;
  margin: auto;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  padding-right: 45px;
}

.c1 {
  border: 1px solid #f1f1f1;
  padding: 28px 0 28px 16px;
}

<body>
  <div>
    <div>
      <div
        class="c0"
      >
        <h1>
          Average price
        </h1>
        <div
          class="c1"
        >
          <div
            class="recharts-responsive-container"
            id="undefined"
            style="width: 100%; height: 300px;"
          >
            <div />
          </div>
          <div
            style="padding-left: 65px;"
          >
            <hr
              class="c2"
            />
            <div
              class="c3"
            >
              <div>
                <div
                  class="c4"
                >
                  <div
                    class="c5"
                    color="#27AE60"
                  />
                  <strong
                    class="c6"
                  >
                    Opus One, Napa Valley, USA
                  </strong>
                  <small
                    class="c7"
                  >
                    Price History
                    e
                  </small>
                </div>
                <div
                  class="c4"
                >
                  <div
                    class="c8"
                    color="#BDBDBD"
                  />
                  <strong
                    class="c6"
                  >
                    Bordeaux Blend Red, Napa Valle
                  </strong>
                  <small
                    class="c7"
                  >
                    Current wine
                  </small>
                </div>
              </div>
              <button
                class="c9"
                style="align-self: flex-end;"
              >
                <div
                  style="margin-right: 10px;"
                />
                 
                Hide bench mark
                k
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
